src/libraries/System.Private.CoreLib/src/System/Buffers/Text/Base64Validator.cs

```diff
@@ -0,0 +1,151 @@
+        /// <param name="decodedLength">The maximum length (in bytes) if you were to decode the base 64 encoded text <paramref name="base64TextUtf8"/> within a byte span.</param>
+        /// <returns>true if <paramref name="base64TextUtf8"/> is decodable; otherwise, false.</returns>
+        public static bool IsValid(ReadOnlySpan<byte> base64TextUtf8, out int decodedLength) =>
+            IsValid<byte, Base64ByteValidatable>(base64TextUtf8, out decodedLength);
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static bool IsValid<T, TBase64Validatable>(ReadOnlySpan<T> base64Text, out int decodedLength)
+            where TBase64Validatable : IBase64Validatable<T>
+        {
+            if (base64Text.IsEmpty)
+            {
+                decodedLength = 0;
+                return true;
+            }
+
+            int length = base64Text.Length;
+            int paddingCount = 0;
+
+            int indexOfPaddingInvalidOrWhitespace = TBase64Validatable.IndexOfAnyExcept(base64Text);
+            if (indexOfPaddingInvalidOrWhitespace >= 0)
+            {
+                while (indexOfPaddingInvalidOrWhitespace >= 0)
+                {
+                    T charToValidate = base64Text[indexOfPaddingInvalidOrWhitespace];
+                    if (TBase64Validatable.IsWhitespace(charToValidate))
+                    {
+                        // Chars to be ignored (e,g, whitespace...) should not count towards the length.
+                        length--;
+                    }
+                    else if (TBase64Validatable.IsEncodingPad(charToValidate))
+                    {
+                        // There can be at most 2 padding chars.
+                        if (paddingCount == 2)
+                        {
+                            decodedLength = 0;
+                            return false;
+                        }
+
+                        paddingCount++;
+                    }
+                    else
+                    {
+                        // An invalid char was encountered.
+                        decodedLength = 0;
+                        return false;
+                    }
+
+                    if (indexOfPaddingInvalidOrWhitespace == base64Text.Length - 1)
+                    {
+                        // The end of the input has been reached.
+                        break;
+                    }
+
+                    // If no padding is found, slice and use IndexOfAnyExcept to look for the next invalid char.
+                    if (paddingCount == 0)
+                    {
+                        ReadOnlySpan<T> slicedSpan = base64Text.Slice(indexOfPaddingInvalidOrWhitespace + 1);
+                        indexOfPaddingInvalidOrWhitespace =
+                            TBase64Validatable.IndexOfAnyExcept(slicedSpan)
+                            + indexOfPaddingInvalidOrWhitespace + 1; // Add current index offset.
```

Comment:
Looks like this method is still missing test coverage. I'm pretty sure input like `" a"` would go into an infinite loop here:
- `indexOfPaddingInvalidOrWhitespace` gets set to 0 because of the initial whitespace
- `paddingCount == 0`, so we get to this branch
- `slicedSpan` = `"a"`
- `IndexOfAnyExcept` returns -1, so you set the index to `-1 + 0 + 1`
- the index is 0 again ... repeat the loop

It may be simpler to write this along the lines of
```c#
int length = 0;

while (true)
{
    int index = IndexOfAnyExcept(base64Text);
    if ((uint)index >= (uint)base64Text.Length)
    {
        length += base64Text.Length;
        break;
    }

    length += index;

    T charToValidate = base64Text[index];
    base64Text = base64Text.Slice(index + 1);

    if (IsWhitespace(charToValidate))
    {
        continue;
    }

    if (!IsEncodingPad(charToValidate))
    {
        return false;
    }

    int paddingCount = 1;
    foreach (T char in base64Text)
    {
        if (IsEncodingPad(char))
        {
            paddingCount++;
        }
        else if (!IsWhitespace(char))
        {
            return false;
        }
    }

    if (paddingCount > 2)
    {
        return false;
    }

    length += paddingCount;
    break;
}

if (length % 4 != 0)
{
    return false;
}
```

Reply:
Apologies I should not have missed that! Thank you for the suggestion, it is much more elegant. I have just made a few slight changes:
- Moved the padding validation outside of the while loop.  
- Return false as soon as more than 2 padding chars detected, instead of counting all then determining if more than 2.  

Please let me know if you preferred it the other way.